### ISSUE TITLE
Support "stay open" mode for Lambda invocations

### ISSUE THREAD

**Issue by jamietanna** [localstack/localstack#4123]
# Type of request: This is a ...

- [ ] bug report
- [x] feature request

# Detailed description

I'm currently using Localstack for local integration + performance tests, to validate the quality of my local Lambda, before it gets shipped to AWS.

Something I've noticed is that, when running Lambdas with languages that perform less well on cold starts (such as Java), the performance of these Lambdas are not ideal.

I'd originally seen this when using the `local` executor, so thought that moving to the `docker` or `docker-reuse` commands would speed this up, but it appears that the container is not actually re-used across invocations, and is deleted each time.

(This may be exacerbated by the fact that we only allow a single Lambda invocation - this may be an additional feature request to add?)

This could be avoided by using the [configuration available in `lambci/lambda`](https://github.com/lambci/docker-lambda#running-in-stay-open-api-mode) to "stay open" allowing the Lambda to stay running, and only be closed when the Localstack container is being closed.

## Expected behavior

Repeat invocations of a Lambda does not need a cold start, as the Lambda will be pre-warmed and re-used.

## Actual behavior

A cold start occurs on each Lambda invocation.

# Steps to reproduce

Localstack version: `0.12.9`

## Command used to start LocalStack

`docker-compose up` via https://gitlab.com/jamietanna/localstack-docker-reuse-not-staying-open/-/blob/main/docker-compose.yml

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

https://gitlab.com/jamietanna/localstack-docker-reuse-not-staying-open

Notice that cold starts occur each time (in the constructor of `Handler`)

**Comment by jamietanna** [comment]
I've had a play with this, and through hacking around things, I've found a solution for this. I'll look at raising a PR soon, at least with the non-test code changes to see if that looks suitable.

My proposal is to have `docker-reuse` use the "stay open" mode, so we can re-use the existing container as per the naming of `reuse` - does that sound OK?

**Comment by whummer** [comment]
HI @jamietanna , thanks for this suggestion. 

> ... `docker-reuse` commands would speed this up, but it appears that the container is not actually re-used across invocations, and is deleted each time

Just to confirm - can you please confirm if that is indeed the case? In fact, the `docker-reuse` Lambda executor should already keep the containers running. 

For example, invoking a Lambda `func1` with `docker-reuse` enabled, the container keeps running, after the invocation is done (for a certain period of time, around ~10 minutes).

```
$ docker ps
CONTAINER ID   IMAGE                         COMMAND                  CREATED         STATUS        PORTS     NAMES
ffbef55ddb7b   localstack/lambda:python3.6   "/private/var/folder…"   2 seconds ago   Up About a minute             localstack_lambda_arn_aws_lambda_us-east-1_000000000000_function_func1
```

Can you reproduce this behavior? Happy to accept a PR to improve the dev experience here - just want to confirm if we're seeing the same behavior with the existing implementation... Thanks!

**Comment by jamietanna** [comment]
Hey, you are correct - I did spot this today but didn't update the issue.

Although the container is reused, the Lambda itself requires a cold start each time.

By adding "stay open" mode we could allow the docker-reuse option to have only one cold start.

I've got this working so can look at raising a PR - hopefully in the next few days.

However, due to the way that the Lambda invocations are locked, a second change would be required to allow separate containers to be spun up, and kept warm (for some time? Indefinitely?) which would allow for a more AWS-like option. I'm happy to raise a second issue to cover that, as that's a bit more work (my initial work on this involves a fair bit more code changing)

**Comment by chaz-doyle-cko** [comment]
@jamietanna - I was about to raise an issue for this as well. We work in .NET and the cold-start time is killing our development time. Your PR & fix would be our saviour!

**Comment by jamietanna** [comment]
Thanks @chaz-doyle-cko - I'll hopefully have it raised by the end of the week

This will only cover leaving it open to avoid cold starts, the concurrent invocations will be a separate PR (which I'll raise a separate issue for)

**Comment by jamietanna** [comment]
Work on this is currently blocked by #4184

**Comment by chaz-doyle-cko** [comment]
Hey @jamietanna, any movement on this? I would offer to help, but I'm not a python guy... :(

**Comment by jamietanna** [comment]
Hey @chaz-doyle-cko unfortunately won't get much time to progress this week. If you're able to give it a play locally with your own lambda to check it works that'd be useful though?

**Comment by chaz-doyle-cko** [comment]
I'll have a look! Will let you know if it works for my stuff... :)

**Comment by barnc** [comment]
Hey @jamietanna - this looks like some amazing work. Just wondering if this issue on hiatus at the moment?

I'm asking as I'm pretty sure we've just run into this same problem (we have a single large jar as a graphql data source) and our init times on localstack are 6000ms+, which is killing us as it's cold starting on every single data fetch.

**Comment by jamietanna** [comment]
Hey, apologies! Unfortunately I've not had much time / motivation to pick this up, as I was having trouble getting the tests passing.

I'd be happy if you wanted to re-implement the fix in #4185 yourself / give me a hand with getting tests passing

**Comment by chaz-doyle-cko** [comment]
Great minds... I've actually got tomorrow set aside to pick this up again... :)

**Comment by barnc** [comment]
Unfortunately I'm in a similar boat - haven't got the python chops to help out meaningfully here I don't think.

Happy to test anything if you think it'd help though.

**Comment by jamietanna** [comment]
For anyone following this, Chaz has picked this up in #4906 :raised_hands:
